File: rustnlpjava/src/main/java/dev/gigapixel/rustnlp/Strr.java
package dev.gigapixel.rustnlp;

import dev.gigapixel.rustnlp.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Strr {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class StrrCleaner implements Runnable {

        MemorySegment segment;
        StrrCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            rustnlp_h.Strr_destroy(this.segment);
        }
    }

    Strr() {}
    Strr(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    
}

File: rustnlpkotlin/src/main/kotlin/dev/gigapixel/rustnlp/Strr.kt
package dev.gigapixel.rustnlp;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure


internal interface StrrLib: Library {
    fun Strr_destroy(handle: Pointer)
}

class Strr internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>
)  {

    internal class StrrCleaner(val handle: Pointer, val lib: StrrLib) : Runnable {
        override fun run() {
            lib.Strr_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<StrrLib> = StrrLib::class.java
        internal val lib: StrrLib = Native.load("rustnlp", libClass)
    }

}
